Automattic/woocommerce-payments
# Determine if the subscription needs to be synced based on the payment date and not the upfront nature of the first payment

Fixes #4565 

#### Changes proposed in this Pull Request

When a merchant selects the **Never (charge the full recurring amount at sign-up)** synchronisation setting, synced subscription products **aren't being synced to their anchor date in Stripe**. 

![Screen Shot 2022-08-11 at 03 15 22](https://user-images.githubusercontent.com/8490476/184043939-b083afe4-9cf9-4ed2-9527-1f5e311f1bb1.png)

After investigating the issue, it turned out this this bug was caused by some logic we had to determine if we should set a trial end date. If the subscription was charged upfront, it wasn't considered synced, however because of this setting, that isn't always the case. This PR fixes this issue by making sure we check if the subscription's first payment date is today when determining if aligning the first payment date using a trial period is necessary. 

#### Testing instructions

* Create a yearly synced subscription synced to any date other than today. 
* Set your Subscription settings to match the screenshot above. 
* Purchase that yearly subscription product. 
    * On `trunk` in the stripe dashboard note that subscription isn't synced - ie it's next payment date will just be 1 year from today.
    * On this branch note that it is synced.
* Test various other sync settings and sync products to ensure there's no regression introduced by this change. 

-------------------


**Post merge**
